### PR TITLE
Migrate to python 3.12

### PR DESCRIPTION
**Pull request by chris48s**
This application is quite similar to https://github.com/DemocracyClub/aggregator-api/ and hence this PR is quite similar to https://github.com/DemocracyClub/aggregator-api/pull/544
I reckon review that one first, and I will annotate this one a bit more lightly. As with https://github.com/DemocracyClub/aggregator-api/pull/544 the core thing is:

- getting onto python 3.12 (and necessary package upgrades to support that)
- linter config (ruff, djhtml)
- aligning CI with other repos

and a few minor bits of faffing with packages to support that.

The code formatting/whitespace changes are in d30c6ef and d888fa8 The other commits are all functional changes.

This application doesn't have a dev environment. I did test a deploy on staging but have now reverted the commits.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,11 +11,18 @@
...
+    * `pip install -r api_endpoints/v1_postcode_lookup/requirements.txt`
 * Run the test suite: `pytest`
 * Run lint checks: `ruff .`
 * Auto-format: `ruff format .`
 
+## Run application
+
+- Frontend: `./mange.py runserver`
+- API: `pipenv run start`
+
```
Comment (chris48s) [inline README.md:25]: In f71864e3 I've done a little bit to make it a bit easier to spin up the application and document some things that I had to work out. I think there are more improvements I would like to make to the way we're handling dependencies and invoking up the application as this is a bit messy. That said, I'm going to stop pulling the thread here and make those things another PR for another day. A sub-optional process which is a bit more clearly documented is at least leaving things a little better than I found them.

File: .circleci/config.yml [hunk 2]
```diff
@@ -7,17 +7,16 @@ orbs:
...
+    - image: cimg/python:3.12
     working_directory: ~/repo
     steps:
     - checkout
     - restore_cache:
         keys:
-        - v3-dependencies-{{ checksum "Pipfile.lock" }}
-        - v3-dependencies-default
+        - v4-dependencies-{{ checksum "Pipfile.lock" }}
+        - v4-dependencies-default
```
Comment (chris48s) [inline .circleci/config.yml:17]: Not really sure how/why this happened, but I somehow saved a bad cache. I have not seen this before.

I'm not going to dwell on this too much - just going to invalidate and move on.

**Comment by chris48s** [comment]
btw, I noticed every time I try to push to this repo, there is a GH workflow that fails to validate with

```
No event triggers defined in `on`
```

I've not attempted to do anything about that here. Any background on what that is? Given it has never run, I'm inclined to bin it.

**Comment by VirginiaDooley** [comment]
> btw, I noticed every time I try to push to this repo, there is a GH workflow that fails to validate with
> 
> ```
> No event triggers defined in `on`
> ```
> 
> I've not attempted to do anything about that here. Any background on what that is? Given it has never run, I'm inclined to bin it.

I'm not sure, sorry.

**Review by VirginiaDooley — APPROVED** [review]

**Pull request merged**